src/main/java/io/appform/dropwizard/actors/base/UnmanagedPublisher.java

```diff
@@ -106,6 +113,79 @@ public final long pendingSidelineMessagesCount() {
     }
 
+    /**
+     * Note: Timeout is in MilliSeconds, Function take a list of message as input and return not ack msg as output
+     * @param messages : Messages to be published
+     * @param properties
+     * @param timeout : in MS timeout for waiting on countDownLatch
+     * @param unit : timeout unit
+     * @return : List of message nacked
+     * @throws Exception
+     */
+    public final List<Message> publishWithConfirmListener(List<Message> messages, AMQP.BasicProperties properties,
+            long timeout, @NotNull TimeUnit unit) throws Exception {
+        publishChannel.confirmSelect();
+        ConcurrentNavigableMap<Long, Message> outstandingConfirms = new ConcurrentSkipListMap<>();
+        final CountDownLatch publishAckLatch = new CountDownLatch(messages.size());
+        publishChannel.addConfirmListener((sequenceNumber, multiple) -> {
+            if (multiple) {
+                ConcurrentNavigableMap<Long, Message> confirmed = outstandingConfirms.headMap(
+                        sequenceNumber, true
+                );
+                for(int i =0;i<confirmed.size();i++)
+                    publishAckLatch.countDown();
+                confirmed.clear();
+            } else {
+                publishAckLatch.countDown();
+                outstandingConfirms.remove(sequenceNumber);
+            }
+            log.info(String.format("Message with delivery tag %s acknoledged", sequenceNumber));
+        }, (sequenceNumber, multiple) -> {
+            {
+                if(multiple == true)
+                {
+                    ConcurrentNavigableMap<Long, Message> confirmed = outstandingConfirms.headMap(
+                            sequenceNumber, true
+                    );
+                    for(int i =0;i<confirmed.size();i++)
+                        publishAckLatch.countDown();
+                }
+                else
+                    publishAckLatch.countDown();
+            }
+            log.debug(String.format("Message nacked with multiple= %s : %s", multiple, outstandingConfirms.get(sequenceNumber)));
+        });
+
+        String routingKey;
+        if (config.isSharded()) {
+            routingKey = NamingUtils.getShardedQueueName(queueName, getShardId());
+        } else {
+            routingKey = queueName;
+        }
+
+        long startTime = System.nanoTime();
+        for (Message message : messages) {
+            try {
+                outstandingConfirms.put(publishChannel.getNextPublishSeqNo(), message);
+                publishChannel.basicPublish(config.getExchange(), routingKey, properties,
+                        mapper().writeValueAsBytes(message));
+            } catch (Exception e) {
+                log.debug(String.format("Failed to publish Message : %s with exception %s", message, e));
+                publishAckLatch.countDown();
+            }
+        }
+
+        if (!publishAckLatch.await(unit.toMillis(timeout), TimeUnit.MILLISECONDS)) {
```

Comment:
Why are we changing this to ms? We can do the await as is right?

Reply:
If someone passes the timeout in any other unit

Reply:
Since you are already taking the `TimeUnit unit` as a param, this conversion is not required. If user wants to wait for 10 mins, pass the time unit as is without converting to millis.